mm2src/coins/eth.rs

```diff
@@ -6179,3 +6190,755 @@ async fn get_eth_gas_details(
+                if amount != &BigDecimal::from(1) {
+                    return Err("ERC-721 amount must be 1".to_string());
+                }
+            },
+        }
+        if taker_secret_hash.len() != 32 {
+            return Err("taker_secret_hash must be 32 bytes".to_string());
+        }
+        if maker_secret_hash.len() != 32 {
+            return Err("maker_secret_hash must be 32 bytes".to_string());
+        }
+
+        Ok(())
+    }
+
+    fn prepare_htlc_data(
+        &self,
+        args: &SendNftMakerPaymentArgs<'_, Self>,
+        taker_address: Address,
+        token_address: Address,
+        time_lock: u32,
+    ) -> Vec<u8> {
+        let id = self.etomic_swap_id(time_lock, args.maker_secret_hash);
+        ethabi::encode(&[
+            Token::FixedBytes(id),
+            Token::Address(taker_address),
+            Token::Address(token_address),
+            Token::FixedBytes(args.taker_secret_hash.to_vec()),
+            Token::FixedBytes(args.maker_secret_hash.to_vec()),
+            Token::Uint(U256::from(time_lock)),
+        ])
+    }
+
+    /// ERC721 contract has overloaded versions of the `safeTransferFrom` function,
+    /// but `Contract::function` method returns only the first if there are overloaded versions of the same function.
+    /// Provided function retrieves the `safeTransferFrom` variant that includes a `bytes` parameter.
+    /// This variant is specifically used for transferring ERC721 tokens with additional data.
+    fn erc721_transfer_with_data(&self) -> Result<&ethabi::Function, Erc721FunctionError> {
+        let functions = ERC721_CONTRACT
+            .functions_by_name("safeTransferFrom")
+            .map_err(|e| Erc721FunctionError::AbiError(ERRL!("{}", e)))?;
+
+        // Find the correct function variant by inspecting the input parameters.
+        let function = functions
+            .iter()
+            .find(|f| {
+                f.inputs.len() == 4
+                    && matches!(
+                        f.inputs.last().map(|input| &input.kind),
+                        Some(&ethabi::ParamType::Bytes)
+                    )
+            })
+            .ok_or_else(|| {
+                Erc721FunctionError::FunctionNotFound(
+                    "Failed to find the correct safeTransferFrom function variant".to_string(),
+                )
+            })?;
+        Ok(function)
+    }
+
+    async fn validate_nft_maker_payment_v2_impl(
+        &self,
+        args: ValidateNftMakerPaymentArgs<'_, Self>,
+    ) -> ValidatePaymentResult<()> {
+        let contract_type = self.parse_contract_type(args.contract_type)?;
+        self.validate_payment_args(
+            args.taker_secret_hash,
+            args.maker_secret_hash,
+            &args.amount,
+            &contract_type,
+        )
+        .map_err(ValidatePaymentError::InternalError)?;
+        let etomic_swap_contract = self.parse_token_contract_address(args.swap_contract_address)?;
```

Comment:
Maybe the function should be renamed to `parse_contract_address` since it's not used for token contracts only.

Reply:
oh, got what have confused you, the name in my mind meant that this function  can parse `token_address` and `contract_address`, where by contract address in swap context we mean etomic swap contract. 
But you are right token_address is actually some smart contract address in blockchain.

fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2084/commits/ba61d495b5db96f79948c6a7714713592f825ca8